src/templates/next/layouts/Content/Content.tsx

```diff
@@ -1,33 +1,140 @@
-import { ContentPageSchema } from "~/engine"
+import { ContentPageSchema, IsomerSitemap } from "~/engine"
 import ContentPageHeader from "../../components/shared/ContentPageHeader"
 import Siderail from "../../components/shared/Siderail"
 import TableOfContents from "../../components/shared/TableOfContents"
 import { Skeleton } from "../Skeleton"
 import { renderComponent } from "../render"
+import { BreadcrumbProps, SiderailProps } from "~/common"
+
+const getBreadcrumbFromSiteMap = (
+  sitemap: IsomerSitemap,
+  permalink: string[],
+): BreadcrumbProps => {
+  const breadcrumb = []
+  let node = sitemap
+  let currentPath = ""
+  for (const pathSegment of permalink) {
+    currentPath += "/" + pathSegment
+    const nextNode = node.children?.find(
+      (node) => node.permalink === currentPath,
+    )
+    if (!nextNode) {
+      // TODO: handle this unexpected case where cannot traverse to permalink in the sitemap
+      break
+    }
+    node = nextNode
+    breadcrumb.push({
+      title: node.title,
+      url: node.permalink,
+    })
+  }
+  return { links: breadcrumb }
+}
+
+const getSiderailFromSiteMap = (
+  sitemap: IsomerSitemap,
+  permalink: string[],
+): SiderailProps | null => {
+  let node = sitemap
+  let currentPath = ""
+
+  let i = 0
+  while (i < permalink.length - 1) {
+    currentPath += "/" + permalink[i]
+    const nextNode = node.children?.find(
+      (node) => node.permalink === currentPath,
+    )
+    if (!nextNode) {
+      // TODO: handle this unexpected case where cannot traverse to parent in the sitemap
+      return null
+    }
+    node = nextNode
+    i++
+  }
+  if (!node.children) {
+    // TODO: handle this unexpected case where parent does not contain current page
+    return null
+  }
+  const parentTitle = node.title
+  const parentUrl = node.permalink
+
+  const pages = []
+  // get all siblings of page
+  const pagePath = "/" + permalink.join("/")
+  for (const sibling of node.children) {
+    if (sibling.permalink === pagePath) {
+      pages.push({
+        title: sibling.title,
+        url: sibling.permalink,
+        isCurrent: true,
+        childPages: sibling.children?.map((child) => ({
+          url: child.permalink,
+          title: child.title,
+        })),
+      })
+    } else {
+      pages.push({
+        title: sibling.title,
+        url: sibling.permalink,
+      })
+    }
+  }
+  return {
+    parentTitle,
+    parentUrl,
+    pages,
+  }
+}
+
+const getTableOfContentsFromContent = ({
+  content,
+}: Pick<ContentPageSchema, "content">) => {
```

Comment:
super nit: can convert this from an object to just directly getting `content` as the first argument (`ContentPageSchema["content"]`)

Reply:
[de186df](https://github.com/isomerpages/isomer-components/pull/62/commits/de186df46d9ffcad61d6bc410c02601ace8e6d0c)